pytest v6.2.0 pytest.approx failing tests when pytest v6.1.2 passes

(This is a cross-post of https://github.com/scikit-hep/pyhf/issues/1219)

- [x] a detailed description of the bug or problem you are having

On 2020-12-12 https://github.com/scikit-hep/pyhf/commit/f824afe77d9e48e90651931700ccfc3d3c268c18 ran on CI on `pyhf`'s `master` and succeeded. Later that day [`pytest` `v6.2.0`](https://github.com/pytest-dev/pytest/releases/tag/6.2.0) was released and the nightly scheduled CI failed on 

```pytb
_______________________ test_optim_with_value[jax-mu=1] ________________________

backend = (<pyhf.tensor.jax_backend.jax_backend object at 0x7f6bf92def50>, None)
source = {'bindata': {'bkg': [100.0, 150.0], 'bkgsys_dn': [98, 100], 'bkgsys_up': [102, 190], 'data': [120.0, 180.0], ...}, 'binning': [2, -0.5, 1.5]}
spec = {'channels': [{'name': 'singlechannel', 'samples': [{'data': [30.0, 95.0], 'modifiers': [{...}], 'name': 'signal'}, {'data': [100.0, 150.0], 'modifiers': [{...}], 'name': 'background'}]}]}
mu = 1.0

    @pytest.mark.parametrize('mu', [1.0], ids=['mu=1'])
    def test_optim_with_value(backend, source, spec, mu):
        pdf = pyhf.Model(spec)
        data = source['bindata']['data'] + pdf.config.auxdata
    
        init_pars = pdf.config.suggested_init()
        par_bounds = pdf.config.suggested_bounds()
    
        optim = pyhf.optimizer
    
        result = optim.minimize(pyhf.infer.mle.twice_nll, data, pdf, init_pars, par_bounds)
        assert pyhf.tensorlib.tolist(result)
    
        result, fitted_val = optim.minimize(
            pyhf.infer.mle.twice_nll,
            data,
            pdf,
            init_pars,
            par_bounds,
            fixed_vals=[(pdf.config.poi_index, mu)],
            return_fitted_val=True,
        )
        assert pyhf.tensorlib.tolist(result)
        assert pyhf.tensorlib.shape(fitted_val) == ()
>       assert pytest.approx(17.52954975, rel=1e-5) == fitted_val
E       assert 17.52954975 ± 1.8e-04 == DeviceArray(17.52954975, dtype=float64)
E        +  where 17.52954975 ± 1.8e-04 = <function approx at 0x7f6cc1747f80>(17.52954975, rel=1e-05)
E        +    where <function approx at 0x7f6cc1747f80> = pytest.approx

tests/test_optim.py:383: AssertionError
```

Diffing the installed libraries between the two (in [f824afe_install.txt](https://github.com/pytest-dev/pytest/files/5684261/f824afe_install.txt) and [failing_install.txt](https://github.com/pytest-dev/pytest/files/5684262/failing_install.txt)) shows that the relevant change is `pytest`

```
$ diff f824afe_install.txt failing_install.txt 
33a34
> importlib-metadata     3.1.1
83c84
< py                     1.9.0
---
> py                     1.10.0
96c97
< pytest                 6.1.2
---
> pytest                 6.2.0
143a145
> zipp                   3.4.0
```

This is confirmed as if

```diff
--- a/setup.py
+++ b/setup.py
@@ -29,7 +29,7 @@
         + extras_require['contrib']
         + extras_require['shellcomplete']
         + [
-            'pytest~=6.0',
+            'pytest~=6.1.0',
             'pytest-cov>=2.5.1',
             'pytest-mock',
             'pytest-benchmark[histogram]',
```

the [CI installs `v6.1.2` and passes](https://github.com/scikit-hep/pyhf/actions/runs/418404132).

The only mention of `pytest.approx`in the [`v6.2.0` release notes](https://github.com/pytest-dev/pytest/releases/tag/6.2.0) is

> 7710: Use strict equality comparison for non-numeric types in pytest.approx instead of
raising TypeError.
>
> This was the undocumented behavior before 3.7, but is now officially a supported feature.

- [x] output of `pip list` from the virtual environment you are using

- [f824afe_install.txt](https://github.com/pytest-dev/pytest/files/5684261/f824afe_install.txt)
- [failing_install.txt](https://github.com/pytest-dev/pytest/files/5684262/failing_install.txt)

- [x] pytest and operating system versions

- Failing: `pytest` `v6.2.0` on `ubuntu-latest` and `macos-latest` on GitHub Actions
- Passing: `pytest` `v6.1.2` on `ubuntu-latest` and `macos-latest` on GitHub Actions

- [x] minimal example if possible

I don't have a trivial minimal failing example at time of reporting (I can work on that if desired), but the CI tests that caught this are detailed in https://github.com/scikit-hep/pyhf/issues/1219 and an example of `v6.1.2` passing can be see in https://github.com/scikit-hep/pyhf/pull/1220.

cc @lukasheinrich @kratsg

--- a/changelog/8132.bugfix.rst
+++ b/changelog/8132.bugfix.rst
@@ -1,0 +1,10 @@
+Fixed regression in ``approx``: in 6.2.0 ``approx`` no longer raises
+``TypeError`` when dealing with non-numeric types, falling back to normal comparison.
+Before 6.2.0, array types like tf.DeviceArray fell through to the scalar case,
+and happened to compare correctly to a scalar if they had only one element.
+After 6.2.0, these types began failing, because they inherited neither from
+standard Python number hierarchy nor from ``numpy.ndarray``.
+
+``approx`` now converts arguments to ``numpy.ndarray`` if they expose the array
+protocol and are not scalars. This treats array-like objects like numpy arrays,
+regardless of size.

--- a/src/_pytest/python_api.py
+++ b/src/_pytest/python_api.py
@@ -543,26 +543,18 @@ def approx(expected, rel=None, abs=None, nan_ok: bool = False) -> ApproxBase:
 
 
 def _is_numpy_array(obj: object) -> bool:
-    """Return true if the given object is implicitly convertible to numpy array.
-
-    A special effort is made to avoid importing numpy unless it's really necessary.
     """
-    import sys
-
-    np: Any = sys.modules.get("numpy")
-    if np is not None:
-        # avoid infinite recursion on numpy scalars, which have __array__
-        if np.isscalar(obj):
-            return False
-        elif isinstance(obj, np.ndarray):
-            return True
-        elif hasattr(obj, "__array__") or hasattr("obj", "__array_interface__"):
-            return True
-    return False
+    Return true if the given object is implicitly convertible to ndarray,
+    and numpy is already imported.
+    """
+    return _as_numpy_array(obj) is not None
 
 
 def _as_numpy_array(obj: object) -> Optional["ndarray"]:
-    """Return an ndarray if obj is implicitly convertible, and numpy is already imported."""
+    """
+    Return an ndarray if the given object is implicitly convertible to ndarray,
+    and numpy is already imported, otherwise None.
+    """
     import sys
 
     np: Any = sys.modules.get("numpy")

--- a/src/_pytest/python_api.py
+++ b/src/_pytest/python_api.py
@@ -15,8 +15,13 @@ from typing import overload
 from typing import Pattern
 from typing import Tuple
 from typing import Type
+from typing import TYPE_CHECKING
 from typing import TypeVar
 from typing import Union
+
+if TYPE_CHECKING:
+    from numpy import ndarray
+
 
 import _pytest._code
 from _pytest.compat import final
@@ -232,10 +237,11 @@ class ApproxScalar(ApproxBase):
     def __eq__(self, actual) -> bool:
         """Return whether the given value is equal to the expected value
         within the pre-specified tolerance."""
-        if _is_numpy_array(actual):
+        asarray = _as_numpy_array(actual)
+        if asarray is not None:
             # Call ``__eq__()`` manually to prevent infinite-recursion with
             # numpy<1.13.  See #3748.
-            return all(self.__eq__(a) for a in actual.flat)
+            return all(self.__eq__(a) for a in asarray.flat)
 
         # Short-circuit exact equality.
         if actual == self.expected:
@@ -521,6 +527,7 @@ def approx(expected, rel=None, abs=None, nan_ok: bool = False) -> ApproxBase:
     elif isinstance(expected, Mapping):
         cls = ApproxMapping
     elif _is_numpy_array(expected):
+        expected = _as_numpy_array(expected)
         cls = ApproxNumpy
     elif (
         isinstance(expected, Iterable)
@@ -536,7 +543,7 @@ def approx(expected, rel=None, abs=None, nan_ok: bool = False) -> ApproxBase:
 
 
 def _is_numpy_array(obj: object) -> bool:
-    """Return true if the given object is a numpy array.
+    """Return true if the given object is implicitly convertible to numpy array.
 
     A special effort is made to avoid importing numpy unless it's really necessary.
     """
@@ -544,8 +551,30 @@ def _is_numpy_array(obj: object) -> bool:
 
     np: Any = sys.modules.get("numpy")
     if np is not None:
-        return isinstance(obj, np.ndarray)
+        # avoid infinite recursion on numpy scalars, which have __array__
+        if np.isscalar(obj):
+            return False
+        elif isinstance(obj, np.ndarray):
+            return True
+        elif hasattr(obj, "__array__") or hasattr("obj", "__array_interface__"):
+            return True
     return False
+
+
+def _as_numpy_array(obj: object) -> Optional["ndarray"]:
+    """Return an ndarray if obj is implicitly convertible, and numpy is already imported."""
+    import sys
+
+    np: Any = sys.modules.get("numpy")
+    if np is not None:
+        # avoid infinite recursion on numpy scalars, which have __array__
+        if np.isscalar(obj):
+            return None
+        elif isinstance(obj, np.ndarray):
+            return obj
+        elif hasattr(obj, "__array__") or hasattr("obj", "__array_interface__"):
+            return np.asarray(obj)
+    return None
 
 
 # builtin pytest.raises helper

--- a/testing/python/approx.py
+++ b/testing/python/approx.py
@@ -447,6 +447,36 @@ class TestApprox:
         assert a12 != approx(a21)
         assert a21 != approx(a12)
 
+    def test_numpy_array_protocol(self):
+        """
+        array-like objects such as tensorflow's DeviceArray are handled like ndarray.
+        See issue #8132
+        """
+        np = pytest.importorskip("numpy")
+
+        class DeviceArray:
+            def __init__(self, value, size):
+                self.value = value
+                self.size = size
+
+            def __array__(self):
+                return self.value * np.ones(self.size)
+
+        class DeviceScalar:
+            def __init__(self, value):
+                self.value = value
+
+            def __array__(self):
+                return np.array(self.value)
+
+        expected = 1
+        actual = 1 + 1e-6
+        assert approx(expected) == DeviceArray(actual, size=1)
+        assert approx(expected) == DeviceArray(actual, size=2)
+        assert approx(expected) == DeviceScalar(actual)
+        assert approx(DeviceScalar(expected)) == actual
+        assert approx(DeviceScalar(expected)) == DeviceScalar(actual)
+
     def test_doctests(self, mocked_doctest_runner) -> None:
         import doctest
 

